wolfcrypt/src/dsa.c

```diff
@@ -97,68 +97,133 @@ void wc_FreeDsaKey(DsaKey* key)
     mp_clear(&key->p);
 }
 
+
+/* validate that (L,N) match allowed sizes from FIPS 186-4, Section 4.2.
+ * l - represents L, the size of p in bits
+ * n - represents N, the size of q in bits
+ * return 0 on success, -1 on error */
+static int CheckDsaLN(int l, int n)
+{
+    int ret = -1;
+
+    switch (l) {
+        case 1024:
+            if (n == 160)
+                ret = 0;
+            break;
+        case 2048:
+            if (n == 224 || n == 256)
+                ret = 0;
+            break;
+        case 3072:
+            if (n == 256)
+                ret = 0;
+            break;
+        default:
+            break;
+    }
+
+    return ret;
+}
+
+
 #ifdef WOLFSSL_KEY_GEN
 
+/* Create DSA key pair (&dsa->x, &dsa->y)
+ *
+ * Based on NIST FIPS 186-4,
+ * "B.1.1 Key Pair Generation Using Extra Random Bits"
+ *
+ * rng - pointer to initialized WC_RNG structure
+ * dsa - pointer to initialized DsaKey structure, will hold generated key
+ *
+ * return 0 on success, negative on error */
 int wc_MakeDsaKey(WC_RNG *rng, DsaKey *dsa)
 {
-    unsigned char *buf;
-    int qsize, err;
+    byte* cBuf;
+    int qSz, pSz, cSz, err;
+    mp_int tmpQ;
 
     if (rng == NULL || dsa == NULL)
         return BAD_FUNC_ARG;
 
-    qsize = mp_unsigned_bin_size(&dsa->q);
-    if (qsize == 0)
+    qSz = mp_unsigned_bin_size(&dsa->q);
+    pSz = mp_unsigned_bin_size(&dsa->p);
+
+    /* verify (L,N) pair bit lengths */
+    if (CheckDsaLN(pSz * 8, qSz * 8) != 0)
         return BAD_FUNC_ARG;
 
-    /* allocate ram */
-    buf = (unsigned char *)XMALLOC(qsize, dsa->heap,
-                                   DYNAMIC_TYPE_TMP_BUFFER);
-    if (buf == NULL)
+    /* generate extra 64 bits so that bias from mod function is negligible */
+    cSz = qSz + 8;
```

Comment:
Isn't 64 / WOLFSSL_BIT_SIZE more clear since your comment says 64 and the code says 8.